subcommand/inject-connect/command.go

```diff
@@ -323,15 +343,73 @@ func (c *Command) Run(args []string) int {
 		TLSConfig: &tls.Config{GetCertificate: c.getCertificate},
 	}
 
-	c.UI.Info(fmt.Sprintf("Listening on %q...", c.flagListen))
-	if err := server.ListenAndServeTLS("", ""); err != nil {
-		c.UI.Error(fmt.Sprintf("Error listening: %s", err))
-		return 1
+	// channel used for health checks
+	// also check to see if we should enable TLS
+	consulAddr := os.Getenv("CONSUL_HTTP_ADDR")
+	if consulAddr == "" {
+		c.UI.Error("CONSUL_HTTP_ADDR is not specified")
+		return 0
+	}
+	consulUrl, err := url.Parse(consulAddr)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error parsing CONSUL_HTTP_ADDR: %s", err))
+		return 0
 	}
 
+	if c.flagEnableHealthChecks {
+		go func() {
+			c.UI.Info(fmt.Sprintf("Listening on %q...", c.flagListen))
+			if err := server.ListenAndServeTLS("", ""); err != nil {
+				c.UI.Error(fmt.Sprintf("Error listening: %s", err))
+			}
+		}()
+
+		reconcilePeriod, err := time.ParseDuration(c.flagConnectInjectHealthCheckReconcilePeriod)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error parsing health-checks-reconcile-period: %s", err))
+			return 1
+		}
+
+		healthResource := connectinject.HealthCheckResource{
+			Log:                 hclog.Default().Named("healthCheckResource"),
+			KubernetesClientset: c.clientset,
+			ConsulUrl:           consulUrl,
+			Ctx:                 ctx,
+			ReconcilePeriod:     reconcilePeriod,
+		}
+
+		ctl := &controller.Controller{
+			Log:      hclog.Default().Named("healthCheckController"),
+			Resource: &healthResource,
+		}
+
+		// Start the health check controller, reconcile is started at the same time
+		// and new events will queue in the informer
+		go func() {
+			ctl.Run(ctx.Done())
```

Comment:
We've added some error handling to this. It's not really possible to test though because it's not possible to make controller exit since it actually has its own panic handling and in general is pretty bomb-proof. We did manually test it by adding a `return` to `ctl.Run()`.